File: FeatureToggle.DAL/DiskFeatureRepository.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using FeatureToggle.Definitions;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Newtonsoft.Json;

namespace FeatureToggle.DAL
{
    public class DiskFeatureRepository : IFeatureRepository
    {
        private readonly string _filepath;
        private readonly List<FeatureConfiguration> _features;
        private readonly ILogger<DiskFeatureRepository> _logger;

        public DiskFeatureRepository(IOptions<FeaturesFileConfiguration> options, ILogger<DiskFeatureRepository> logger)
        {
            _logger = logger;
            _filepath = options.Value.FeaturesConfigurationFile;
            _features = LoadConfigurationFile();
        }

        private List<FeatureConfiguration> LoadConfigurationFile()
        {
            if (_filepath == null || !File.Exists(_filepath))
            {
                _logger.LogDebug($"File <{_filepath}> does not exist.");
                return new List<FeatureConfiguration>();
            }

            var json = File.ReadAllText(_filepath);
            return JsonConvert.DeserializeObject<List<FeatureConfiguration>>(json);
        }

        private void WriteConfigurationFile()
        {
            var json = JsonConvert.SerializeObject(_features);
            File.WriteAllText(_filepath, json);
        }

        public IEnumerable<FeatureConfiguration> Select(string pattern)
        {
            if (string.IsNullOrEmpty(pattern))
                return _features;

            return _features.Where(pair => pair.Feature.StartsWith(pattern, StringComparison.InvariantCultureIgnoreCase));
        }

        public void Delete(string featureName)
        {
            if (_features.RemoveAll(configuration => string.Equals(configuration.Feature, featureName, StringComparison.InvariantCultureIgnoreCase)) == 0)
                throw new KeyNotFoundException($"Feature <{featureName}> is not configured.");

            WriteConfigurationFile();
        }

        public IEnumerable<FeatureConfiguration> Update(string featureName, string newValue)
        {
            if (string.IsNullOrWhiteSpace(featureName))
                throw new ArgumentException("Feature name cannot be empty", nameof(featureName));

            var feature = _features.Find(feat =>
                feat.Feature.Equals(featureName, StringComparison.InvariantCultureIgnoreCase));

            if (feature == null)
                throw new KeyNotFoundException($"Feature <{featureName}> is not configured.");

            feature.Value = newValue;

            WriteConfigurationFile();

            return _features;
        }

        public void Add(string featureName, string newValue)
        {
            if (string.IsNullOrWhiteSpace(featureName))
                throw new InvalidDataException("parameter <featureName> cannot be empty");

            _features.Add(new FeatureConfiguration(featureName, newValue));

            WriteConfigurationFile();
        }
    }
}

File: F#/FeatureToggle.DiskRepository/DiskFeatureRepository.fs
namespace FeatureToggle.DAL

open System
open System.Collections.Generic
open FeatureToggle.Definitions
open DiskStorage

module DiskFeatureRepository =

    let private splitWhen pred =
        let rec loop acc = function
            | [] -> None
            | x::xs when pred x -> Some (List.rev acc, x::xs)
            | x::xs -> loop (x::acc) xs
        loop []

    let private matches fname = (fun feature -> feature.Feature = fname)

    let private IRepository_Select (storage:DiskStorage) pattern = 
        let features = storage.readFile()
        match pattern with
            | null | "" -> features
            | ptrn -> features |> List.filter (fun p -> p.Feature.StartsWith( ptrn, StringComparison.InvariantCultureIgnoreCase))
    
    let private IRepository_Delete storage featureName =
        let features = storage.readFile()
        
        let split = features |> splitWhen (matches featureName)
        let newFeatures = match split with
                            | None -> raise (KeyNotFoundException "FeatureName not found")
                            | Some (first, last) -> List.concat [first; List.tail last]

        storage.writeFile newFeatures
        newFeatures

    let private IRepository_Set storage featureName newValue =
        match featureName with
        | "" | null -> raise (ArgumentException "FeatureName cannot be empty")
        | _ ->  
            let features = storage.readFile()
            let split = features |> splitWhen (matches featureName)
            let newFeatures = match split with
                                | None -> {Feature = featureName; Value = newValue}::features
                                | Some (first, head::tail) -> List.concat [first; [ {head with Value = newValue}]; tail]
            storage.writeFile newFeatures
            newFeatures

    let createRepository storage = {
        Select = IRepository_Select storage
        Delete = IRepository_Delete storage
        Set = IRepository_Set storage
    }
